### PR TITLE
Add separate thread to writes files

### PR DESCRIPTION
**Pull request by smu160**
## Overview: What does this pull request change?

Add a thread-safe queue and a listener thread that takes frames and outputs them to a file.

## Motivation and Explanation: Why and how do your changes improve the library?

Since optimizations are [heavily encouraged](https://docs.manim.community/en/stable/contributing/performance.html) by the community, the changes here improve performance by up to 15% (in limited benchmarks/use cases).

Steps to reproduce the results are as follows:

Example script:
```python
import time

from manim import *

class MovingFrameBox(Scene):
    def construct(self):
        start = time.time_ns()
        text=MathTex(
            "\\frac{d}{dx}f(x)g(x)=","f(x)\\frac{d}{dx}g(x)","+",
            "g(x)\\frac{d}{dx}f(x)"
        )
        self.play(Write(text))
        framebox1 = SurroundingRectangle(text[1], buff = .1)
        framebox2 = SurroundingRectangle(text[3], buff = .1)
        self.play(
            Create(framebox1),
        )
        self.wait()
        self.play(
            ReplacementTransform(framebox1,framebox2),
        )
        self.wait()
        end = time.time_ns()
        print(f"{end - start} ns")
```

Run
```bash
manim -p example.py MovingFrameBox
```

Machine info:
```bash
CPU: Apple M2
MEM: 24 GB
OS: macOS 14.1 arm64
Kernel: 23.1.0
```

## Further Information and Comments

Since concurrency can introduce a myriad of issues, this PR will need to be thoroughly reviewed. Please let me know if any additional information is needed.

Thank you!

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Comment by smu160** [comment]
@MrDiver Apologies for the multiple PR requests. I closed out the previous PR, #3270 when I synced my fork with the main repo. I will continue to work on that PR here.

Thank you!

**Comment by behackl** [comment]
@smu160, just to clarify: are you still working on this PR, or is this ready to be looked at? I'll convert it to a draft if it is not ready for review yet.

**Comment by smu160** [comment]
@behackl  l'm still actively working on this PR, and I plan to address any remaining issues shortly. I'll notify you once it's ready for review. In the interim, if you have any concerns or suggestions, feel free to let me know. 

Thank you!

**Comment by MrDiver** [comment]
Heyo, sorry for the long time to reply, no worries if you just scrapped the old PR, i am fine with closing PRs that are rewritten if they're not really used anymore, so if this PR is ready we might take a look at it in the next weeks.

Can you shortly elaborate what things were changed ?

+ If you like there is something called [Spin](https://spinroot.com/spin/whatispin.html) it's a formal verification tool for parallel applications. I learned that at university and never really had a use case for it by now, but maybe now is the time to put that into practice xD

I just leave the link here if you're interested, maybe you also find a more modern tool for this! (But we actually do not need to verify that the code is 100% correct xD just as a fun thing to do)

**Comment by MrDiver** [comment]
Conlficting ?  #3501

File: manim/utils/hashing.py [hunk 1]
```diff
@@ -74,7 +74,7 @@ def layer(func):
             if is_method:
                 return lambda self, obj: cls._handle_already_processed(
                     obj,
-                    default_function=lambda obj: func(self, obj),
```
Comment (MrDiver) [inline manim/utils/hashing.py:77]: keep

File: manim/utils/hashing.py [hunk 2]
```diff
@@ -214,29 +214,35 @@ def default(self, obj: Any):
                 # See https://github.com/ManimCommunity/manim/pull/402.
                 code = ""
             return self._cleaned_iterable({"code": code, "nonlocals": cvardict})
-        elif isinstance(obj, np.ndarray):
+        if isinstance(obj, np.ndarray):
             if obj.size > 1000:
                 obj = np.resize(obj, (100, 100))
-                return f"TRUNCATED ARRAY: {repr(obj)}"
+                return f"TRUNCATED ARRAY: {obj.tobytes()}"
             # We return the repr and not a list to avoid the JsonEncoder to iterate over it.
```
Comment (MrDiver) [inline manim/utils/hashing.py:221]: Comment still valid ?

File: manim/scene/scene.py [hunk 3]
```diff
@@ -1264,19 +1265,19 @@ def check_interactive_embed_is_valid(self):
                 "Disabling interactive embed as 'skip_animation_preview' is enabled",
             )
             return False
-        elif config["write_to_movie"]:
```
Comment (MrDiver) [inline manim/scene/scene.py:1267]: unrelated changes ?

**Review by MrDiver — CHANGES_REQUESTED** [review]
You made a lot of changes in the PR which are likely unrelated to the actual change with adding a separate Thread to render the files.
Maybe it would be more useful to have them in another PR because it makes it really hard to see what the actual changes are with this PR and what is just unrelated behavior change.

File: manim/scene/scene.py (hunk 3 above)
Comment (smu160) [reply manim/scene/scene.py:1267]: Yes, I figured that I may as well add small touch ups and fixes as I go through all the other changes I made. I'm happy to separate this into another PR if you'd prefer. 

These changes are just to bring the code base in line with [PEP8](https://peps.python.org/pep-0008/). 

From the aforementioned link, consider:

```python
For sequences, (strings, lists, tuples), use the fact that empty sequences are false:

# Correct:
if not seq:
if seq:

# Wrong:
if len(seq):
if not len(seq):
```

For a mathematical perspective, we can think of non-empty sequences as an equivalence class. Similarly, the non-zero values are an equivalence class.

File: manim/utils/hashing.py (hunk 1 above)
Comment (smu160) [reply manim/utils/hashing.py:77]: This change was suggested to my by pylint. I'm happy to change the naming back.

File: manim/utils/hashing.py (hunk 2 above)
Comment (smu160) [reply manim/utils/hashing.py:221]: I will remove this comment

**Comment by smu160** [comment]
Hi @MrDiver,

First and foremost, I apologize for the delayed response -- I've been heads down at work. I really appreciate your patience.

Thank you for thoroughly reviewing the pull request. I understand your concern regarding the changes in this PR that are unrelated to adding a separate writer thread. I think segregating the unrelated changes into separate PR(s) will help with clarifying all the changes made here. I have made the necessary adjustments, and I will push up the changes shortly.

In the interim, I'd like to explain all the changes I made:

- Use a separate thread to write to files
    - Redesign for concurrency [^1].
        - For the purposes of for communication between threads, we slightly modify the design. The basic idea here is to change the structure of the `SceneFIleWriter` such that a thread-safe queue is put in between FFMPEG and the process that runs the `SceneFileWriter`. Namely, `SceneFileWriter` is only respondible adding the frames to the queue.
    - Get parallelism for free [^1]
        - If we design for concurrency, then we can get parallelism for free (assuming we have more threads to throw at the problem). Thus, we add a separate thread by utilizing the new `listen_and_write()` method. This method is launched as a separate thread. The thread continuously checks the queue for new frames, and writes them to the file -- this frees the main thread to continue other work.
    - Cleanup
        - Once the `SceneFileWriter` calls either `finish()` or `close_movie_pipe()`, we cleanup the thread by signaling that we are finished. We do this by sending a `-1` to the queue. The `listen_and_write` thread always checks if it receives a `-1`, and stops upon receipt of this value.

Given this change, I don't think [#3501](https://github.com/ManimCommunity/manim/pull/3501) is a conflict. On the contrary, the new design will mesh well with that change 

Other changes related to performance

- Encode numpy array(s) as bytes
    - Converting a numpy array to bytes is faster than representing it as a string. I am happy to provide concrete figures if needed
- Sort `animations_list` in-place
    - this way an extra allocation is avoided
- Unroll the loop when applying `zlib.crc32`
    - The amount of items to loop through is small and constant, so we can remove the overhead of the loop

Thank you and Happy New Year!
        
[^1]: [Concurrency is not Parallelism by Rob Pike](https://www.youtube.com/watch?v=oV9rvDllKEg).
